### ISSUE TITLE
Changes to TravisCI / Move CI to Github Actions

### ISSUE THREAD

**Issue by JoaoRodrigues** [biopython/biopython#3340]
TravisCI sent out an email this week announcing changes to their plans ([link](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing)). The post is quite dense and it's hard to understand if this affects us but there seem to be new limitations to open-source projects. Also, we use the travis-ci.org portal, which will be closed down at the end of this year ([link](https://docs.travis-ci.com/user/migrate/open-source-repository-migration#frequently-asked-questions)) so we might take this opportunity to think about updating our CI infrastructure a little.

Currently, we use TravisCI and Appveyor to test on Windows and Linux, and it seems we don't test on MacOS. Moving to Github Actions would let us test on all three major OSes and a bunch of Python versions including PyPy. It would also allow us to run up to 256 jobs in parallel, and in my experience, it seems those jobs run faster than on Travis/Appveyor.

To kickstart this discussion and test drive a possible transition to Github Actions, I started a branch (purposedly out of date with the current master by a few commits) implementing the style checking, packaging, and simple testing steps. You can see the branch [here](https://github.com/JoaoRodrigues/biopython/tree/gh_actions_tests) and the file I have been working on [here](https://github.com/JoaoRodrigues/biopython/blob/gh_actions_tests/.github/workflows/basic.yml).

Thoughts? Suggestions?

**Comment by peterjc** [comment]
If this is the end of free TravisCI testing (which as Joao says is hard to understand from the post), then so be it. We can be grateful for the years of free support they have given us. We could explore using OBF funds to pay for TravisCI, but speaking as current OBF president I would rather Biopython explore other free alternatives first.

Personally I also have experience using CircleCI, which should be a fairly simple change - the configuration file is different, but the approach is quite similar apart from the lack of TravisCI stages. They use the same business model TravisCI did, free for open source, paid for closed projects. I could have a go at that?

In parallel, GitHub Actions sounds like a good long term alternative - and may assist with more interactive bots on our pull requests too in future? I'd be delighted for Joao to take the lead on that.

**Comment by JoaoRodrigues** [comment]
>If this is the end of free TravisCI testing (which as Joao says is hard to understand from the post), then so be it. We can be grateful for the years of free support they have given us. We could explore using OBF funds to pay for TravisCI, but speaking as current OBF president I would rather Biopython explore other free alternatives first.

I agree that exploring free options is better. Personally, I think OBF funds should be spent on people, not machines :)

>Personally I also have experience using CircleCI, which should be a fairly simple change - the configuration file is different, but the approach is quite similar apart from the lack of TravisCI stages. They use the same business model TravisCI did, free for open source, paid for closed projects. I could have a go at that?

Sounds good! I had a look at their webpage but couldn't figure out if they support all 3 OSes in their free/open-source plans.

>In parallel, GitHub Actions sounds like a good long term alternative - and may assist with more interactive bots on our pull requests too in future? I'd be delighted for Joao to take the lead on that.

Happy to work on it. We can also brainstorm changes to the testing strategy. For example, we could run different sets of the test suite in parallel (online, offline), add more services, etc. Are there any ideas from previous discussions on this topic?
